db/migrations/57_add_expires_at_into_groups_groups_and_groups_ancestors.sql

```diff
@@ -0,0 +1,128 @@
+-- +migrate Up
+ALTER TABLE `groups_groups` ADD COLUMN `expires_at` datetime  NOT NULL DEFAULT '9999-12-31 23:59:59'
+    COMMENT 'If set, the group membership expires at the specified time';
+ALTER TABLE `groups_ancestors` ADD COLUMN `expires_at` datetime NOT NULL DEFAULT '9999-12-31 23:59:59'
+    COMMENT 'If set, the group relation expires at the specified time';
+DROP TRIGGER `before_update_groups_groups`;
+-- +migrate StatementBegin
+CREATE TRIGGER `before_update_groups_groups` BEFORE UPDATE ON `groups_groups` FOR EACH ROW BEGIN
+    IF (OLD.child_group_id != NEW.child_group_id OR OLD.parent_group_id != NEW.parent_group_id OR OLD.type != NEW.type OR NOT OLD.expires_at <=> NEW.expires_at) THEN
+        INSERT IGNORE INTO `groups_propagate` (`id`, `ancestors_computation_state`) VALUES (OLD.child_group_id, 'todo')
+        ON DUPLICATE KEY UPDATE `ancestors_computation_state` = 'todo';
+        INSERT IGNORE INTO `groups_propagate` (`id`, `ancestors_computation_state`) (
+            SELECT `groups_ancestors`.`child_group_id`, 'todo'
+            FROM `groups_ancestors`
+            WHERE `groups_ancestors`.`ancestor_group_id` = OLD.`child_group_id`
+        ) ON DUPLICATE KEY UPDATE `ancestors_computation_state` = 'todo';
+        DELETE `bridges` FROM `groups_ancestors` `child_descendants`
+            JOIN `groups_ancestors` `parent_ancestors`
+            JOIN `groups_ancestors` `bridges`
+                ON (`bridges`.`ancestor_group_id` = `parent_ancestors`.`ancestor_group_id` AND
+                    `bridges`.`child_group_id` = `child_descendants`.`child_group_id`)
+        WHERE `parent_ancestors`.`child_group_id` = OLD.`parent_group_id` AND
+            `child_descendants`.`ancestor_group_id` = OLD.`child_group_id`;
+
+        INSERT IGNORE INTO `groups_propagate` (id, ancestors_computation_state) VALUES (NEW.child_group_id, 'todo')
+        ON DUPLICATE KEY UPDATE `ancestors_computation_state` = 'todo';
+    END IF;
+END
+-- +migrate StatementEnd
+DROP TRIGGER `before_delete_groups_groups`;
+-- +migrate StatementBegin
+CREATE DEFINER=`algorea`@`%` TRIGGER `before_delete_groups_groups` BEFORE DELETE ON `groups_groups` FOR EACH ROW BEGIN
```

Comment:
2 INSERT and 4 DELETE at each update/delete. 😲
Have you tried deciphering if everything is still needed? I didn't. (you can answer "no", that's just a question)

Reply:
As you can see there had been 3 INSERTs + 4 DELETEs at each update and 2 INSERTs + 4 DELETES at each deletion. I got rid of some unnecessary queries and only 3 INSERTs + 1 DELETE at each update and 2 INSERTs + 1 DELETE at each deletion remained.